### PR TITLE
fix: error when there is __init__.py in the migration folder

### PR DESCRIPTION
**Pull request by LanceMoe**

### CODE REVIEW

**Comment by LanceMoe** [comment]
Aerich changed from `sql` to `py` source code. When there is `__init__.py` in the migration folder, aerich will report an error. This PR fixes the issue.

File: aerich/migrate.py [hunk 1]
```diff
@@ -50,7 +50,7 @@ class Migrate:
     @classmethod
     def get_all_version_files(cls) -> List[str]:
         return sorted(
-            filter(lambda x: x.endswith("py"), os.listdir(cls.migrate_location)),
+            filter(lambda x: x.endswith("py") and not x.startswith("__"), os.listdir(cls.migrate_location)),
```
Comment (baseplate-admin) [inline aerich/migrate.py]: Instead of this approach. Why dont we do `x.startswith('__init__')`.

Comment (LanceMoe) [reply aerich/migrate.py]: Thanks for your review, I have edited it, please check again.

**Review by baseplate-admin — APPROVED** [review]
私のアドバイスを聞いてくれてありがとう

**Review by waketzheng — CHANGES_REQUESTED** [review]
`key=lambda x: int(x.split("_")[0])` show that if there is any file whose name does not startswith a number will raise a ValueError.
So this function can be refactor to be:
```py
def get_all_version_files(cls) -> List[str]:
        pattern = re.compile(r"(\d+)_")
        files = [i.name for i in cls.migrate_location.glob("*.py") if pattern.match(i.name)]
        return sorted(files, key=lambda x: int(pattern.search(x).group(1)))
```

**Comment by LanceMoe** [comment]
@waketzheng 
Based on your code, I commit a code that can pass mypy. Please review again, thank you.

File: aerich/migrate.py [hunk 2]
```diff
@@ -54,10 +55,16 @@ def get_field_by_name(name: str, fields: List[dict]) -> dict:
...
+        pattern = re.compile(r"(\d+)_")
+        files = [i.name for i in cls.migrate_location.glob("*.py") if pattern.match(i.name)]
+        
+        def sort_key(x: str):
+            match = pattern.search(x)
+            if match:
+                return int(match.group(1))
+            raise ValueError(f"Invalid version file name: {x}")
+        
+        return sorted(files, key=sort_key)
```
Comment (waketzheng) [inline aerich/migrate.py]: Too complex, use `return sorted(files, key=lambda x: int(x.split("_")[0]))` instead

File: aerich/migrate.py [hunk 3]
```diff
@@ -54,10 +55,16 @@ def get_field_by_name(name: str, fields: List[dict]) -> dict:
 
     @classmethod
     def get_all_version_files(cls) -> List[str]:
-        return sorted(
-            filter(lambda x: x.endswith("py"), os.listdir(cls.migrate_location)),
-            key=lambda x: int(x.split("_")[0]),
-        )
+        pattern = re.compile(r"(\d+)_")
+        files = [i.name for i in cls.migrate_location.glob("*.py") if pattern.match(i.name)]
```
Comment (waketzheng) [inline aerich/migrate.py]: Let's make it more simple: `files = [p.name for p in cls.migrate_location.glob("*.py") if p.name[0].isdigit()]`

**Comment by waketzheng** [comment]
@LanceMoe This PR is useful. It allow user to add custom file like `utils.py` to the migration directory.

Code can be as simple as:
```py
    def get_all_version_files(cls) -> List[str]:
        files = [p.name for p in cls.migrate_location.glob("*.py") if p.name[0].isdigit()]
        return sorted(files, key=lambda x: int(x.split("_")[0]))
```
Note: `import re` should be removed, and would be better if you can add some test cases.

**Comment by LanceMoe** [comment]
@waketzheng 
Thanks for your suggestion, but I think this will cause some potential bugs. For example
```
if p.name[0].isdigit()
```
If `p.name` is `'123foo'`, `p.name[0].isdigit()` will be true, but `int(x.split("_")[0])` will raise an error.
I'll see how to improve it.

**Comment by LanceMoe** [comment]
@waketzheng 
Following your idea, I submitted another change and added tests.

Finally, I chose to add two closures in this function for the following reasons:
1. I think that although it can be combined into one line, but it will make the rules of version files unclear. Through the `is_version_file` function, we can clearly know what rules a migration file should follow: It is a `.py` file, and the file name contains an underscore, and the underscore in the file name is preceded by a number.
2. Although `split("_")[0]` is not long, it is reused. Wrapping it into a function can not only improve the readability of the code, but also prevent mistakes during refactoring of feature changes in the future. Although this function has only one line, the reason why I did not use lambda to define it, is that it is not convenient to add type hints when lambda is defined alone. This requires importing `Callable` from the `typing` library. I think this loses the meaning of using lambda. It is better to define it as a normal closure function.
3. Although the type of `migrate_location` is marked as a `Path`, the existing test code mocks `os.listdir`, and `migrate_location` is directly passed a `str` value, so it becomes inconvenient to use `glob("*.py")` I chose to switch back to `filter`+`os.listdir` so that I don't have to modify the test code a lot. If refactoring is needed, I think it should be opened as a separate task instead of solving it in this PR.

Please review again, thank you.

**Comment by waketzheng** [comment]
Looks nice, just change to this:
```diff
     @classmethod
     def get_all_version_files(cls) -> List[str]:
-        def get_file_version(file_name: str):
+        def get_file_version(file_name: str) -> str:
             return file_name.split("_")[0]

-        def is_version_file(file_name: str):
+        def is_version_file(file_name: str) -> bool:
             if not file_name.endswith("py"):
                 return False
-            if file_name.find("_") <= 0:
+            if "_" not in file_name:
                 return False
             return get_file_version(file_name).isdigit()
```

**Comment by LanceMoe** [comment]
@waketzheng 
Modified and committed, please check again, thank you!

**Review by waketzheng — APPROVED** [review]

**Pull request merged**